File: server/app/generators/mines_gen.py
"""
Mines (Solitaire / Paper Minesweeper) Generator and Solver
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import random
from typing import List, Tuple, Optional, Dict, Any, Set

class MinesGenerator:
    """
    Procedural generator for Solitaire Paper Minesweeper (Simon Tatham mines.c / puzzle-magazine.com).
    Guarantees 100% solvability by pure human deduction with zero guessing.
    """

    DIFFICULTY_SETTINGS = {
        "easy": {"rows": 8, "cols": 8, "mines": 8, "target_clues": 28},
        "medium": {"rows": 8, "cols": 8, "mines": 12, "target_clues": 22},
        "hard": {"rows": 8, "cols": 8, "mines": 15, "target_clues": 17},
    }

    def __init__(self, seed: Optional[int] = None):
        if seed is not None:
            random.seed(seed)

    def _get_neighbors(self, r: int, c: int, rows: int, cols: int) -> List[Tuple[int, int]]:
        neighbors = []
        for dr in (-1, 0, 1):
            for dc in (-1, 0, 1):
                if dr == 0 and dc == 0:
                    continue
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols:
                    neighbors.append((nr, nc))
        return neighbors

    def solve_deductive(self, rows: int, cols: int, clues: Dict[Tuple[int, int], int], total_mines: int) -> Tuple[bool, List[List[int]]]:
        """
        Solves using pure deductive logic:
        - Rule 1: Direct saturation (remaining unknown neighbors == remaining mines needed -> all mines)
        - Rule 2: Direct clearing (remaining mines needed == 0 -> all remaining unknown neighbors are safe)
        - Rule 3: Subset difference logic (U_A subset of U_B)
        - Rule 4: Global mine count capacity
        Returns (is_solved, state_grid where -1=unknown, 0=safe, 1=mine).
        """
        state = [[-1] * cols for _ in range(rows)]
        # Clue cells are known safe
        for (r, c) in clues:
            state[r][c] = 0

        changed = True
        while changed:
            changed = False

            # Rules 1 & 2: Local clue saturation & clearing
            for (r, c), val in clues.items():
                neighs = self._get_neighbors(r, c, rows, cols)
                unknowns = [p for p in neighs if state[p[0]][p[1]] == -1]
                mines = [p for p in neighs if state[p[0]][p[1]] == 1]
                rem = val - len(mines)

                if rem < 0 or len(unknowns) + len(mines) < val:
                    return False, state  # Contradiction

                if rem == 0 and unknowns:
                    for ur, uc in unknowns:
                        state[ur][uc] = 0
                        changed = True
                elif rem == len(unknowns) and unknowns:
                    for ur, uc in unknowns:
                        state[ur][uc] = 1
                        changed = True

            if changed:
                continue

            # Rule 3: Subset difference logic between clue pairs
            clue_list = list(clues.items())
            for i in range(len(clue_list)):
                (r1, c1), v1 = clue_list[i]
                n1 = self._get_neighbors(r1, c1, rows, cols)
                u1 = set(p for p in n1 if state[p[0]][p[1]] == -1)
                m1 = sum(1 for p in n1 if state[p[0]][p[1]] == 1)
                rem1 = v1 - m1

                for j in range(len(clue_list)):
                    if i == j:
                        continue
                    (r2, c2), v2 = clue_list[j]
                    n2 = self._get_neighbors(r2, c2, rows, cols)
                    u2 = set(p for p in n2 if state[p[0]][p[1]] == -1)
                    m2 = sum(1 for p in n2 if state[p[0]][p[1]] == 1)
                    rem2 = v2 - m2

                    if u1 and u1.issubset(u2):
                        diff = u2 - u1
                        rem_diff = rem2 - rem1
                        if rem_diff < 0:
                            return False, state
                        if rem_diff == 0 and diff:
                            for dr, dc in diff:
                                state[dr][dc] = 0
                                changed = True
                        elif len(diff) == rem_diff and diff:
                            for dr, dc in diff:
                                state[dr][dc] = 1
                                changed = True
                if changed:
                    break

            if changed:
                continue

            # Rule 4: Global mine capacity
            found_mines = sum(1 for r in range(rows) for c in range(cols) if state[r][c] == 1)
            all_unknowns = [(r, c) for r in range(rows) for c in range(cols) if state[r][c] == -1]
            rem_global = total_mines - found_mines

            if rem_global == 0 and all_unknowns:
                for ur, uc in all_unknowns:
                    state[ur][uc] = 0
                    changed = True
            elif len(all_unknowns) == rem_global and all_unknowns:
                for ur, uc in all_unknowns:
                    state[ur][uc] = 1
                    changed = True

        is_solved = all(state[r][c] != -1 for r in range(rows) for c in range(cols))
        return is_solved, state

    def generate(self, difficulty: str = "medium") -> Dict[str, Any]:
        difficulty = difficulty.lower()
        cfg = self.DIFFICULTY_SETTINGS.get(difficulty, self.DIFFICULTY_SETTINGS["medium"])
        rows = cfg["rows"]
        cols = cfg["cols"]
        num_mines = cfg["mines"]
        target_clues = cfg["target_clues"]

        all_cells = [(r, c) for r in range(rows) for c in range(cols)]

        for _ in range(100):
            mine_set = set(random.sample(all_cells, num_mines))

            # Compute full clues for all non-mine cells
            full_clues: Dict[Tuple[int, int], int] = {}
            for r in range(rows):
                for c in range(cols):
                    if (r, c) not in mine_set:
                        neighs = self._get_neighbors(r, c, rows, cols)
                        full_clues[(r, c)] = sum(1 for p in neighs if p in mine_set)

            # Check if fully solvable with all non-mine clues revealed
            is_solved, _ = self.solve_deductive(rows, cols, full_clues, num_mines)
            if not is_solved:
                continue

            # Iteratively prune clues down to target while preserving solvability
            pruned_clues = dict(full_clues)
            clue_keys = list(full_clues.keys())
            random.shuffle(clue_keys)

            for k in clue_keys:
                if len(pruned_clues) <= target_clues:
                    break
                val = pruned_clues.pop(k)
                solved, _ = self.solve_deductive(rows, cols, pruned_clues, num_mines)
                if not solved:
                    pruned_clues[k] = val  # Restore

            # Build grid for representation (-1 = blank unrevealed, 0..8 = clue number)
            puzzle_grid = [[-1] * cols for _ in range(rows)]
            for (r, c), val in pruned_clues.items():
                puzzle_grid[r][c] = val

            solution_grid = [[0] * cols for _ in range(rows)]
            for r, c in mine_set:
                solution_grid[r][c] = 1

            mine_coords = sorted(list(mine_set))

            return {
                "title": "MINES",
                "difficulty": difficulty.upper(),
                "rows": rows,
                "cols": cols,
                "total_mines": num_mines,
                "clues_count": len(pruned_clues),
                "puzzle": puzzle_grid,
                "solution": solution_grid,
                "mine_coords": mine_coords
            }

        raise RuntimeError("Failed to generate deducible Mines puzzle")

    def format_ascii(self, puzzle_data: Dict[str, Any]) -> str:
        rows = puzzle_data["rows"]
        cols = puzzle_data["cols"]
        puzzle = puzzle_data["puzzle"]
        lines = []
        lines.append("--- MINES ---")
        lines.append(f"DIFFICULTY: {puzzle_data['difficulty']}")
        total_mines = puzzle_data.get('total_mines', 12)
        lines.append(f"Use the numbered clues showing adjacent mine counts to deduce each of the {total_mines} hidden mines across the grid.")
        lines.append(f"TOTAL MINES: {puzzle_data['total_mines']}")
        lines.append("")

        sep = "   +" + "---+" * cols
        lines.append(sep)
        for r in range(rows):
            line = "   |"
            for c in range(cols):
                val = puzzle[r][c]
                line += "   |" if val == -1 else f" {val} |"
            lines.append(line)
            lines.append(sep)

        return "\n".join(lines)

File: esp32-firmware/src/generators/MinesGen.cpp
#include "MinesGen.h"
#include "../printer/EscPosPrinter.h"
#include <string.h>

static void shuffleCoords(uint8_t* arr, uint8_t n) {
    for (int i = n - 1; i > 0; i--) {
        int j = random(0, i + 1);
        uint8_t temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

static uint8_t getNeighbors(uint8_t r, uint8_t c, uint8_t nRow[8], uint8_t nCol[8]) {
    uint8_t count = 0;
    for (int dr = -1; dr <= 1; dr++) {
        for (int dc = -1; dc <= 1; dc++) {
            if (dr == 0 && dc == 0) continue;
            int nr = r + dr;
            int nc = c + dc;
            if (nr >= 0 && nr < 8 && nc >= 0 && nc < 8) {
                nRow[count] = nr;
                nCol[count] = nc;
                count++;
            }
        }
    }
    return count;
}

MinesGen::MinesGen() : _totalMines(12), _cluesCount(0) {
    memset(_puzzle, -1, sizeof(_puzzle));
    memset(_solution, 0, sizeof(_solution));
}

bool MinesGen::solveDeductive(int8_t puz[8][8], uint8_t totalMines) {
    int8_t state[8][8];
    memset(state, -1, sizeof(state));

    // Clue cells are known safe
    for (uint8_t r = 0; r < 8; r++) {
        for (uint8_t c = 0; c < 8; c++) {
            if (puz[r][c] != -1) {
                state[r][c] = 0;
            }
        }
    }

    bool changed = true;
    while (changed) {
        changed = false;

        // Rules 1 & 2: Local saturation & clearing
        for (uint8_t r = 0; r < 8; r++) {
            for (uint8_t c = 0; c < 8; c++) {
                int8_t val = puz[r][c];
                if (val == -1) continue;

                uint8_t nRow[8], nCol[8];
                uint8_t nCount = getNeighbors(r, c, nRow, nCol);

                uint8_t uCount = 0;
                uint8_t mCount = 0;
                uint8_t uRow[8], uCol[8];

                for (uint8_t i = 0; i < nCount; i++) {
                    int8_t st = state[nRow[i]][nCol[i]];
                    if (st == -1) {
                        uRow[uCount] = nRow[i];
                        uCol[uCount] = nCol[i];
                        uCount++;
                    } else if (st == 1) {
                        mCount++;
                    }
                }

                int rem = val - mCount;
                if (rem < 0 || (uCount + mCount) < val) {
                    return false; // Contradiction
                }

                if (rem == 0 && uCount > 0) {
                    for (uint8_t i = 0; i < uCount; i++) {
                        state[uRow[i]][uCol[i]] = 0;
                        changed = true;
                    }
                } else if (rem == uCount && uCount > 0) {
                    for (uint8_t i = 0; i < uCount; i++) {
                        state[uRow[i]][uCol[i]] = 1;
                        changed = true;
                    }
                }
            }
        }

        if (changed) continue;

        // Rule 3: Subset difference logic between adjacent clue pairs
        for (uint8_t r1 = 0; r1 < 8; r1++) {
            for (uint8_t c1 = 0; c1 < 8; c1++) {
                int8_t v1 = puz[r1][c1];
                if (v1 == -1) continue;

                uint8_t nRow1[8], nCol1[8];
                uint8_t nCount1 = getNeighbors(r1, c1, nRow1, nCol1);
                uint8_t uCount1 = 0, mCount1 = 0;
                uint8_t u1[8]; // indices encoded as r * 8 + c
                for (uint8_t i = 0; i < nCount1; i++) {
                    if (state[nRow1[i]][nCol1[i]] == -1) u1[uCount1++] = nRow1[i] * 8 + nCol1[i];
                    else if (state[nRow1[i]][nCol1[i]] == 1) mCount1++;
                }
                if (uCount1 == 0) continue;
                int rem1 = v1 - mCount1;

                // Compare against other clues within 2 steps
                for (int dr = -2; dr <= 2; dr++) {
                    for (int dc = -2; dc <= 2; dc++) {
                        if (dr == 0 && dc == 0) continue;
                        int r2 = r1 + dr;
                        int c2 = c1 + dc;
                        if (r2 < 0 || r2 >= 8 || c2 < 0 || c2 >= 8) continue;
                        int8_t v2 = puz[r2][c2];
                        if (v2 == -1) continue;

                        uint8_t nRow2[8], nCol2[8];
                        uint8_t nCount2 = getNeighbors(r2, c2, nRow2, nCol2);
                        uint8_t uCount2 = 0, mCount2 = 0;
                        uint8_t u2[8];
                        for (uint8_t i = 0; i < nCount2; i++) {
                            if (state[nRow2[i]][nCol2[i]] == -1) u2[uCount2++] = nRow2[i] * 8 + nCol2[i];
                            else if (state[nRow2[i]][nCol2[i]] == 1) mCount2++;
                        }
                        if (uCount2 == 0) continue;
                        int rem2 = v2 - mCount2;

                        // Check if u1 is subset of u2
                        bool isSubset = true;
                        for (uint8_t a = 0; a < uCount1; a++) {
                            bool found = false;
                            for (uint8_t b = 0; b < uCount2; b++) {
                                if (u1[a] == u2[b]) { found = true; break; }
                            }
                            if (!found) { isSubset = false; break; }
                        }

                        if (isSubset) {
                            uint8_t diff[8];
                            uint8_t diffCount = 0;
                            for (uint8_t b = 0; b < uCount2; b++) {
                                bool inU1 = false;
                                for (uint8_t a = 0; a < uCount1; a++) {
                                    if (u2[b] == u1[a]) { inU1 = true; break; }
                                }
                                if (!inU1) diff[diffCount++] = u2[b];
                            }

                            int remDiff = rem2 - rem1;
                            if (remDiff < 0) return false;
                            if (remDiff == 0 && diffCount > 0) {
                                for (uint8_t d = 0; d < diffCount; d++) {
                                    state[diff[d] / 8][diff[d] % 8] = 0;
                                    changed = true;
                                }
                            } else if (remDiff == diffCount && diffCount > 0) {
                                for (uint8_t d = 0; d < diffCount; d++) {
                                    state[diff[d] / 8][diff[d] % 8] = 1;
                                    changed = true;
                                }
                            }
                        }
                    }
                }
            }
        }

        if (changed) continue;

        // Rule 4: Global mine count
        uint8_t foundMines = 0;
        uint8_t totalUnknowns = 0;
        uint8_t uCoords[64];
        for (uint8_t r = 0; r < 8; r++) {
            for (uint8_t c = 0; c < 8; c++) {
                if (state[r][c] == 1) foundMines++;
                else if (state[r][c] == -1) uCoords[totalUnknowns++] = r * 8 + c;
            }
        }

        int remGlobal = totalMines - foundMines;
        if (remGlobal == 0 && totalUnknowns > 0) {
            for (uint8_t i = 0; i < totalUnknowns; i++) {
                state[uCoords[i] / 8][uCoords[i] % 8] = 0;
                changed = true;
            }
        } else if (remGlobal == totalUnknowns && totalUnknowns > 0) {
            for (uint8_t i = 0; i < totalUnknowns; i++) {
                state[uCoords[i] / 8][uCoords[i] % 8] = 1;
                changed = true;
            }
        }
    }

    for (uint8_t r = 0; r < 8; r++) {
        for (uint8_t c = 0; c < 8; c++) {
            if (state[r][c] == -1) return false;
        }
    }
    return true;
}

void MinesGen::generate(MinesDifficulty difficulty) {
    uint8_t targetClues = 22;
    if (difficulty == MINES_EASY) {
        _totalMines = 8;
        targetClues = 28;
    } else if (difficulty == MINES_MEDIUM) {
        _totalMines = 12;
        targetClues = 22;
    } else {
        _totalMines = 15;
        targetClues = 17;
    }

    uint8_t allCoords[64];
    for (uint8_t i = 0; i < 64; i++) allCoords[i] = i;

    for (uint8_t attempt = 0; attempt < 50; attempt++) {
        shuffleCoords(allCoords, 64);
        memset(_solution, 0, sizeof(_solution));
        for (uint8_t i = 0; i < _totalMines; i++) {
            _solution[allCoords[i] / 8][allCoords[i] % 8] = 1;
        }

        // Full clues for all non-mine cells
        int8_t fullClues[8][8];
        memset(fullClues, -1, sizeof(fullClues));
        for (uint8_t r = 0; r < 8; r++) {
            for (uint8_t c = 0; c < 8; c++) {
                if (_solution[r][c] == 0) {
                    uint8_t nRow[8], nCol[8];
                    uint8_t nCount = getNeighbors(r, c, nRow, nCol);
                    uint8_t mineCount = 0;
                    for (uint8_t i = 0; i < nCount; i++) {
                        if (_solution[nRow[i]][nCol[i]] == 1) mineCount++;
                    }
                    fullClues[r][c] = mineCount;
                }
            }
        }

        if (!solveDeductive(fullClues, _totalMines)) {
            continue;
        }

        // Prune clues down to targetClues
        memcpy(_puzzle, fullClues, sizeof(_puzzle));
        uint8_t nonMineCoords[64];
        uint8_t nonMineCount = 0;
        for (uint8_t i = _totalMines; i < 64; i++) {
            nonMineCoords[nonMineCount++] = allCoords[i];
        }
        shuffleCoords(nonMineCoords, nonMineCount);

        _cluesCount = nonMineCount;
        for (uint8_t i = 0; i < nonMineCount; i++) {
            if (_cluesCount <= targetClues) break;
            uint8_t r = nonMineCoords[i] / 8;
            uint8_t c = nonMineCoords[i] % 8;
            int8_t saved = _puzzle[r][c];
            _puzzle[r][c] = -1;

            if (solveDeductive(_puzzle, _totalMines)) {
                _cluesCount--;
            } else {
                _puzzle[r][c] = saved;
            }
        }

        return; // Successfully generated!
    }
}

void MinesGen::printToReceipt(EscPosPrinter& printer, MinesDifficulty diff) {
    const char* diffStr = (diff == MINES_EASY) ? "EASY" : ((diff == MINES_HARD) ? "HARD" : "MEDIUM");

    printer.setBold(true);
    printer.println("--- MINES ---");
    printer.setBold(false);
    printer.println(String("DIFFICULTY: ") + diffStr);
    printer.println("Use the numbered clues showing adjacent mine counts");
    printer.println("to deduce and mark every hidden mine across the grid.");
    printer.println(String("TOTAL MINES: ") + String(_totalMines));
    printer.println("");

    String sep = "   +---+---+---+---+---+---+---+---+";
    printer.println(sep);

    for (uint8_t r = 0; r < 8; r++) {
        String line = "   |";
        for (uint8_t c = 0; c < 8; c++) {
            int8_t val = _puzzle[r][c];
            if (val == -1) line += "   |";
            else line += String(" ") + val + " |";
        }
        printer.println(line);
        printer.println(sep);
    }
    printer.println("");
}
